### PR TITLE
Fix #28 Adding base user

### LINKED ISSUE #28: Add base user
Need to add a base user from which the Seeker and Owner will "inherit" (or take the fields of the base user - not sure if inherit is the right term for Django models but thats the idea)

### PR DESCRIPTION
**Pull request by Micha546**
Adding a users app to the project,
Adding user functionality to the admin page,
Adding a user model which inherits from the Django class `AbstractBaseUser`,
and uses `PermissionsMixin` for the authentication of the user
Adding basic tests for the `create_superuser` and `create_user` functions

fix #28

### CODE REVIEW

**Comment by AmitAharoni** [comment]
Good job!

**Comment by Micha546** [comment]
fix #28

**Comment by Omeramsc** [comment]
Please add the `fix` comment to the main message, 
otherwise, it'll not link the issue to the PR (look at `Linked issues` on the right side of the screen), meaning It'll be harder to track AND the ticket will need to be closed manually when this PR is done. 

Also, please describe in the description what does this PR include.

File: users/models.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                                        BaseUserManager)
+
+
+class UserManager(BaseUserManager):
+
+    def create_user(self, email, first_name,
+                    last_name, birth_date, password,
+                    **other_fields):
+
+        if not email:
```
Comment (simzacks) [inline users/models.py:14]: You set required fields in the model so why have separate if statements here?

File: users/models.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+        return user
+
+    def create_superuser(self, email, first_name,
+                         last_name, birth_date, password,
+                         **other_fields):
+
+        other_fields.setdefault('is_staff', True)
+        other_fields.setdefault('is_superuser', True)
+
+        if other_fields.get('is_staff') is not True:
```
Comment (simzacks) [inline users/models.py]: instead of "is not True", you might want to use `== False`, or even better:
`if not other_fields.get('is_staff') :`
and when you use `get` you can define a default, so instead of above:
`other_fields.setdefault('is_staff', True)`
you can have a single statement:
`if not other_fields.get('is_staff', True) :`

File: users/models.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+class User(AbstractBaseUser, PermissionsMixin):
+
+    email = models.EmailField(unique=True)
+    first_name = models.CharField(max_length=50)
+    last_name = models.CharField(max_length=50)
+    date_joined = models.DateTimeField(default=timezone.now)
+    birth_date = models.DateField(auto_now=False, auto_now_add=False)
+    is_staff = models.BooleanField(default=False)
+    is_active = models.BooleanField(default=True)
+
```
Comment (simzacks) [inline users/models.py:62]: You don't have an "is_superuser" field

**Comment by Micha546** [comment]
Done, 
I still can't see the issue in the `linked issues` though

File: users/migrations/0001_initial.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+from django.db import migrations, models
+import django.utils.timezone
+
+
+class Migration(migrations.Migration):
+
+    initial = True
+
+    dependencies = [
+        ('auth', '0012_alter_user_first_name_max_length'),
```
Comment (simzacks) [inline users/migrations/0001_initial.py:12]: what is this dependency ?

**Review by simzacks — CHANGES_REQUESTED** [review]
Aside from my comments, my assumption is that you tested it and it works.
It would also be nice to see a test function that verifies that your functions work that we can add to the CI (workflows)

File: users/models.py (hunk 1 above)
Comment (Micha546) [reply users/models.py:14]: From what I read about REQUIRED_FIELDS, it's only used to set which fields the "create_superuser" function will ask for,
So you still have to check the fields in the "create_user" function.

Some references:
https://github.com/django/django/blob/main/docs/topics/auth/customizing.txt - lines 569 to 594
https://docs.djangoproject.com/en/3.1/topics/auth/customizing/

Maybe I'm getting it wrong, ill wait for your reply before changing those lines.

Comment (Micha546) [reply users/models.py:14]: Or do you mean because blank=False is the default model option, there's no need to check them in "create_user"?

File: users/models.py (hunk 2 above)
Comment (Micha546) [reply users/models.py]: I've changed the if statement to `if not other_fields.get('is_staff')`.

But I had to keep `other_fields.setdefault` because the `get` method does not change the vlaue of the key,
it's only a return parameter that the `get` method returns if it doesn't find the key.

source:
https://www.programiz.com/python-programming/methods/dictionary/get

File: users/models.py (hunk 3 above)
Comment (Micha546) [reply users/models.py:62]: Added the is_superuser field.

File: users/migrations/0001_initial.py (hunk 4 above)
Comment (Micha546) [reply users/migrations/0001_initial.py:12]: From what I understand, this dependency has something to do with the `PermissionsMixin` class.
I think that there's a many to many connection somewhere there with `django.auth`,
so `migrate` has to run on django.auth first, and this is why theres the dependency.

"Since `User` inherits from `AbstractUser`, it will inherit `user_permissions` which is a many-to-many field with `auth.Permission`. (It will also inherit a many-to-many field with `auth.Group`)."
an answer from "tim-mccurrach" to a question posted on
https://stackoverflow.com/questions/65540414/django-claiming-inconsistent-migration-history-on-first-migration
(In our case we inherit `user_permissions` from `PermissionsMixin`)

PermissionsMixin class for refrence:
https://docs.djangoproject.com/en/1.8/_modules/django/contrib/auth/models/

File: users/models.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        if not other_fields.get('is_superuser'):
+            raise ValueError('superuser must be assigned to is_superuser=True')
+
+        return self.create_user(email, first_name,
+                                last_name, birth_date, password,
+                                **other_fields)
+
+
+class User(AbstractBaseUser, PermissionsMixin):
```
Comment (nadavsu) [inline users/models.py:52]: I think the `tags` field we discussed should go here too, no?

Comment (Micha546) [reply users/models.py:52]: Isn't it an apartment feature?
I thought we will add it in the apartment profile after this PR marges.
What do you think?

Comment (nadavsu) [reply users/models.py:52]: Each tag is an interest/hobby of an apartment owner or seeker, so it should go be user field in my opinion.
The tags will just be a set of strings. We'll use the sets to calculate a matching score between two users.

File: users/migrations/0001_initial.py (hunk 4 above)
Comment (Micha546) [reply users/migrations/0001_initial.py:12]: Another source I found:
https://stackoverflow.com/questions/51084875/why-does-initial-migration-have-dependencies

File: users/models.py (hunk 5 above)
Comment (DanielMy777) [reply users/models.py:52]: I think its more of a user feature too, but both ways are supposed to work.

Comment (AmitAharoni) [reply users/models.py:52]: IMO it should be User fields.
cuz both Seeker and Owner have them.

 Great job!

Comment (Micha546) [reply users/models.py:52]: Ok, ill add `tags` to the User. But because it's not a built-in field like others, I think it will be best if we merged this first and then add the `tags` field later this week, this way we can work on models that rely on the BaseUser simultaneously.

**Review by DanielMy777 — COMMENTED** [review]
Very nice work @Micha546 !
All additions look fitting to me

File: users/migrations/0001_initial.py (hunk 4 above)
Comment (simzacks) [reply users/migrations/0001_initial.py:12]: Part of the initial stack overflow question was: "but I am now having to remove that dependency from the user migration manually before it will migrate."
This is because it will look for that file if anyone clones the project and tries to migrate. Since the file 0012_alter_user_first_name_max_length doesn't exist, it will cause the migrate to fail.

File: users/models.py (hunk 1 above)
Comment (simzacks) [reply users/models.py:14]: You are correct about the REQUIRED_FIELDS.
What you can do is add null=true and blank=true to any required fields in the model, which will then prevent you from adding them in. This is preferred so that you don't have to check if the required fields are populated each time you change the model.
https://docs.djangoproject.com/en/3.1/ref/models/fields/

**Review by simzacks — CHANGES_REQUESTED** [review]

File: users/migrations/0001_initial.py (hunk 4 above)
Comment (Micha546) [reply users/migrations/0001_initial.py:12]: I think he had another problem, maybe he somehow altered the data-base, and from what I read I think you have to have this dependency: (tl;dr in the bottom)

In the question he says that he gets a problem because "`Migration users.0001_initial is applied before its dependency auth.0012_alter_user_first_name_max_length on database 'default'.`" - not beacuse migrate doesn't find 0012_alter_user_first_name_max_length.
"tim-mccurrach" says it happent because of: `InconsistentMigrationHistory` error, and says: "I don't know how you have deleted your database, but clearly this table (`django_migrations`) hasn't been deleted properly, or it's been meddled with since being wiped."
He also says in the "Can I just delete the problematic dependency?" section of his answer: "No, this will probably cause you problems. Since `User` inherits from `AbstractUser`, it will inherit `user_permissions` which is a many-to-many field with `auth.Permission`. (It will also inherit a many-to-many field with `auth.Group`)."

In the second link I sent, "kr85" asks why his custom user model has the dependency: "0009_alter_user_last_name_max_length" (It's a pretty old post from 2 years ago), and "solarissmoke" answers: "This is normal and expected behaviour - you still need the migrations from `django.contrib.auth` to be applied to your database."

And I have found the file in: `PATH_TO_DJANGO/contrib/auth/migrations` (for me PATH_TO_DJANGO is `/home/vagrant/.local/share/virtualenvs/vagrant-gKDsaKU3/lib/python3.8/site-packages/django`).

tl;dr: I think you have to have this dependency, and the file does exist in `PATH_TO_DJANGO/contrib/auth/migrations` - I'll wait for your reply - maybe I'm getting everything here wrong!

Comment (simzacks) [reply users/migrations/0001_initial.py:12]: I'll approve it if you tell me that if someone clones the repo and runs 'migrate' that it will work without an error

File: users/models.py (hunk 1 above)
Comment (Micha546) [reply users/models.py:14]: If I'll set the fields to blank=true, won't it make it so they won't be required when creating a user? I read in the link you sent that `blank=true` will make it so the field won't be required in a from, do you want the fields to be non-mandatory?

File: users/migrations/0001_initial.py (hunk 4 above)
Comment (Micha546) [reply users/migrations/0001_initial.py:12]: @tamirho tried to clone the repo from my fork yesterday and ran `vagrant up` (which ran `migrate` beacuse of `bootstrap.sh`) and successfully created a superuser, logged into the admin page and ran tests.

Comment (DanielMy777) [reply users/migrations/0001_initial.py:12]: I can confirm this too, just tried it and it works without an error. (Added screenshots to the slack channel)

File: users/models.py (hunk 1 above)
Comment (simzacks) [reply users/models.py:14]: My mistake.  I meant null=False, blank=False.

**Review by simzacks — APPROVED** [review]

**Comment by simzacks** [comment]
@Omeramsc @1Oreo please review

File: users/tests.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                )
+
+        # Check if create_superuser must have is_superuser=True
+        with self.assertRaises(ValueError):
+            with transaction.atomic():
+                db.objects.create_superuser(
+                    'email_2@address.com', 'first_name', 'last_name', '1900-01-01', 'password', is_superuser=False
+                )
+
+    def test_user(self):
```
Comment (Omeramsc) [inline users/tests.py:46]: Usually, when you have a lot of assertions in a test, It's a sign that you need to divide it to separate test functions. 
instead of adding comments before each assertion, have a separate named test for each.

for example: 
```
test_email_not_unique_should_raise():
    with self.assertRaises(IntegrityError):
        with transaction.atomic():
            db.objects.create_user(
                'email@address.com', 'first_name', 'last_name', '1900-01-01', 'password'
            )
```

if some of the tests have something in common they need in order to work (like the user model), maybe it can be added as a fixture
https://docs.pytest.org/en/stable/fixture.html

Comment (Omeramsc) [reply users/tests.py:46]: Another option can be using pytest's `parametrize` - this way you can run the same test with different parameters on each run.
in this case, the parameters can be the values you insert to the DB (either as a single parameter, or separately), and the specific error that should be raised. 

https://docs.pytest.org/en/stable/parametrize.html

----------------------------------------------------------------------
Both are very useful and powerful tools pytest provides, and you can write much better tests if you'll get used to them.

**Review by Omeramsc — CHANGES_REQUESTED** [review]

File: users/models.py (hunk 1 above)
Comment (Micha546) [reply users/models.py:14]: Hey @simzacks, should I get rid of the if statements in the create_user and create_superuser altogether? the reason I added them in the first place is because I was following a tutorial video on how to set up a custom user and in the video the commentator recommended adding if statements to later have easier time writing tests for the user. If it's better to have no if statements, I'll remove them. but then what exception should I expect in the tests?

File: users/models.py (hunk 5 above)
Comment (Micha546) [reply users/models.py:52]: We agreed that @nadavsu will take care of the `tags` in the base user after the merging of this PR, marking this as "resolved".

**Review by Omeramsc — APPROVED** [review]
LGTM
